pontoon/static/js/tags/admin/table.js

```diff
@@ -0,0 +1,182 @@
+
+import React from 'react';
+import ReactTable from 'react-table';
+import "react-table/react-table.css";
+
+import Checkbox from 'widgets/checkbox';
+
+
+export default class TagResourceTable extends React.Component {
+
+    constructor (props) {
+        super(props);
+        this.visible = [];
+        this.state = {checked: new Set()};
+        this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
+        this.handleSubmit = this.handleSubmit.bind(this);
+        this.handleTableChange = this.handleTableChange.bind(this);
+        this.handleTableSortChange = this.handleTableSortChange.bind(this);
+        this.handleTableResize = this.handleTableResize.bind(this);
+        this.renderResource = this.renderResource.bind(this);
+        this.renderResourceCheckbox = this.renderResourceCheckbox.bind(this);
+        this.handleSelectAll = this.handleSelectAll.bind(this);
+        this.renderSelectAllCheckbox = this.renderSelectAllCheckbox.bind(this);
+    }
+
+    get columns () {
+        // column descriptors, as expected by ReactTable
+        return [
+            {Header: this.renderSelectAllCheckbox,
+             Cell: this.renderResourceCheckbox,
+             sortable: false,
+             width: 45
+            },
+            {Header: "Resource",
+             id: "type",
+             Cell: this.renderResource
+            }];
+    }
+
+    get defaultPageSize () {
+        return 5;
+    }
+
+    get selected () {
+        // get the pruned list of selected resources, returns all=true/false
+        // if all visible resources are selected
+        const checked = this.pruneChecked();
+        let all = false;
+        // some rows can be empty strings if there are more visible rows than resources
+        if (checked.size > 0 && checked.size === this.visible.filter(v => v !== '').length) {
+            all = true;
+        }
+        return {all: all, checked: checked};
+    }
+
+    componentWillUpdate (prevProps) {
+        if (prevProps.data !== this.props.data) {
+            const {checked} = this.state;
+            this.visible.length = 0;
+            checked.clear();
```

Comment:
intentionally so

if you call setState in all of these places it recurses - simply emptying the set works

Reply:
Well doesn't that imply there's an underlying architectural problem then? Isn't there another way of writing this so that we can use `setState` to change the state everywhere?

Reply:
i dont see that tbh - i think what is happening here is we are saying - we want to empty (etc) that set, but we dont want to force a refresh

Reply:
ok, i went through code and found that it was either already setting state or was ok to wherever needed - ive updated so that it never touches the copy of the checked set itself